File: csharp/helper_classes/RngExtensions.cs
using Godot;

namespace MazePerformanceGrade.csharp.helper_classes;

public static class RngExtensions
{
    /// <summary>
    /// generate a random vector between 0,0 and max, exclusive
    /// </summary>
    /// <param name="size">the exclusive max value</param>
    /// <returns></returns>
    public static Vector2I RandomVector(this RandomNumberGenerator rng, Vector2I size)
    {
        return new Vector2I(rng.RandiRange(0, size.X - 1), rng.RandiRange(0, size.Y - 1));
    }

    public static void ShuffleNaive<T>(this RandomNumberGenerator rng, T[] array)
    {
        for (int i = 0; i < array.Length; i++)
        {
            var swap = rng.RandiRange(i, array.Length - 1);
            (array[swap], array[i]) = (array[i], array[swap]);
        }
    }
}

File: rust/src/maze_scripts/helper_classes/rng_extensions.rs
use godot::engine::RandomNumberGenerator;
use godot::prelude::*;

pub trait RngExtensions {
    fn random_vector(&mut self, size: Vector2i) -> Vector2i;
    fn shuffle_naive<T>(&mut self, arr: &mut [T]);
}

impl RngExtensions for RandomNumberGenerator {
    fn random_vector(&mut self, size: Vector2i) -> Vector2i {
        let x = self.randi_range(0, size.x - 1);
        let y = self.randi_range(0, size.y - 1);
        Vector2i { x, y }
    }

    fn shuffle_naive<T>(&mut self, arr: &mut [T]) {
        for i in 0..arr.len() {
            let swap = self.randi_range(i as i32, (arr.len() - 1) as i32) as usize;
            arr.swap(i, swap);
        }
    }
}